File: tracker_benchmark/util/calcSeqErrRobust.m
function [aveErrCoverage, aveErrCenter,errCoverage, errCenter] = calcSeqErrRobust(results, rect_anno)

% LineWidth = 2;
% LineStyle = '-';%':';%':' '.-'

% lostCount = zeros(length(seqs), length(trks));
% thred = 0.33;
% 
% errCenterAll=[];
% errCoverageAll=[];

seq_length = results.len;

if strcmp(results.type,'rect')
    for i = 2:seq_length
        r = results.res(i,:);
        r_anno = rect_anno(i,:);
        if (isnan(r) | r(3)<=0 | r(4)<=0)&(~isnan(r_anno))
            results.res(i,:)=results.res(i-1,:);
%             results.res(i,:) = [1,1,1,1];
        end
    end
end

% lenTotalSeq = lenTotalSeq + seq_length;

centerGT = [rect_anno(:,1)+(rect_anno(:,3)-1)/2 rect_anno(:,2)+(rect_anno(:,4)-1)/2];

rectMat = zeros(seq_length, 4);
switch results.type
    case 'rect'
        rectMat = results.res;
    case 'ivtAff'
        for i = 1:seq_length
            [rect c] = calcRectCenter(results.tmplsize, results.res(i,:));
            rectMat(i,:) = rect;
            %                     center(i,:) = c;
        end
    case 'L1Aff'
        for i = 1:seq_length
            [rect c] = calcCenter_L1(results.res(i,:), results.tmplsize);
            rectMat(i,:) = rect;
        end
    case 'LK_Aff'
        for i = 1:seq_length
            [corner c] = getLKcorner(results.res(2*i-1:2*i,:), results.tmplsize);
            rectMat(i,:) = corner2rect(corner);
        end
    case '4corner'
        for i = 1:seq_length
            rectMat(i,:) = corner2rect(results.res(2*i-1:2*i,:));
        end
    case 'affine'
        for i = 1:seq_length
            rectMat(i,:) = corner2rect(results.res(2*i-1:2*i,:));
        end
    case 'SIMILARITY'
        for i = 1:seq_length
            warp_p = parameters_to_projective_matrix(results.type,results.res(i,:));
            [corner c] = getLKcorner(warp_p, results.tmplsize);
%             [corner c] = getORIAcorner(results.res(i,:), results.tmplsize, results.type);
            rectMat(i,:) = corner2rect(corner);
        end
end

% if isfield(results,'shiftType')
%     center = [rectMat(:,1)+(rectMat(:,3)-1)/2 rectMat(:,2)+(rectMat(:,4)-1)/2];
%     ratio=1;
% 
%     shiftType=results.shiftType;
%     switch shiftType
%         case 'scale_8'
%             ratio=0.8;
%         case 'scale_9'
%             ratio=0.9;
%         case 'scale_11'
%             ratio=1.1;
%         case 'scale_12'
%             ratio=1.2;
%     end
% 
%     w = rectMat(:,3)/ratio;
%     h = rectMat(:,4)/ratio;
%     rectMat = round([center(:,1)-w/2,center(:,2)-h/2,w,h]);  
% end

rectMat(1,:) = rect_anno(1,:);

center = [rectMat(:,1)+(rectMat(:,3)-1)/2 rectMat(:,2)+(rectMat(:,4)-1)/2];

errCenter = sqrt(sum(((center(1:seq_length,:) - centerGT(1:seq_length,:)).^2),2));

index = rect_anno>0;
idx=(sum(index,2)==4);
% errCoverage = calcRectInt(rectMat(1:seq_length,:),rect_anno(1:seq_length,:));
tmp = calcRectInt(rectMat(idx,:),rect_anno(idx,:));

errCoverage=-ones(length(idx),1);
errCoverage(idx) = tmp;
errCenter(~idx)=-1;

aveErrCoverage = sum(errCoverage(idx))/length(idx);

aveErrCenter = sum(errCenter(idx))/length(idx);

File: butil/calc_seq_err_robust.py
import butil
import math

def calc_seq_err_robust(results, rect_anno):
    seq_length = int(results['len'])
    res = results['res']
    centerGT = [[r[0]+(r[2]-1)/2.0, r[1]+(r[3]-1)/2.0] for r in rect_anno]
    
    rectMat = [[0, 0, 0, 0]] * seq_length

    resultType = results['type']
    if resultType == 'rect':
        rectMat = res
    elif resultType == 'ivtAff':
        for i in range(seq_length):
            # rect, c, corn = butil.calc_rect_center(results['tmplsize'], res[i])
            rect = butil.rect_affine_IVT(results['tmplsize'][0], res[i])
            rectMat[i] = rect
    elif resultType == 'L1Aff':
        for i in range(seq_length):
            # rect, c = butil.calc_center_L1(res[i], results['tmplsize'])
            rect = butil.rect_affine_L1(results['tmplsize'][0], res[i])
            rectMat[i] = rect
    elif resultType == 'LK_Aff':
        for i in range(seq_length):
            rect = butil.rect_affine_LK(results['tmplsize'][0], 
                res[2*i:2*(i+1)])
            rectMat[i] = rect
    elif resultType == '4corner':
        for i in range(seq_length):
            # corner = res[2*i:2*(i+1)]
            # rectMat[i] = butil.d_to_f(m.corenr2rect(corner, nargout=1)[0])
            rect = butil.rect_4corners(res[2*i:2*(i+1)])
            rectMat[i] = rect
    elif resultType == 'affine':
        for i in range(seq_length):
            rect = butil.rect_4corners(res[2*i:2*(i+1)])
            rectMat[i] = rect
    elif resultType == 'SIMILARITY':
        for i in range(seq_length):
            rect = butil.rect_similarity(results['tmplsize'][0], res[i])
            rectMat[i] = rect
            # wapr_p = m.parameters_to_projective_matrix(resultType, res[i],
            #     nargout=1)
            # corenr, c = m.getLKcorner(wapr_p, results['tmplsize'], nargout=2)
            # rectMat[i] = butil.do_to_f(m.corner2rect(corner, nargout=1)[0])

    rectMat[0] = rect_anno[0]
    center = [[r[0]+(r[2]-1)/2.0, r[1]+(r[3]-1)/2.0] for r in rectMat]
    errCenter = [round(butil.ssd(center[i], centerGT[i]),4)
        for i in range(seq_length)]

    idx = [sum([x>0 for x in r])==4 for r in rect_anno]
    tmp = calc_rect_int(rectMat, rect_anno)

    errCoverage = [-1] * seq_length
    totalerrCoverage = 0
    totalerrCenter = 0

    for i in range(seq_length):
        if idx[i]:
            errCoverage[i] = tmp[i]
            totalerrCoverage += errCoverage[i]
            totalerrCenter += errCenter[i]
        else:
            errCenter[i] = -1

    aveErrCoverage = totalerrCoverage / float(sum(idx))
    aveErrCenter = totalerrCenter / float(sum(idx))

    return aveErrCoverage, aveErrCenter, errCoverage, errCenter

def calc_rect_int(A, B):
    leftA = [a[0] for a in A]
    bottomA = [a[1] for a in A]
    rightA = [leftA[i] + A[i][2] - 1 for i in range(len(A))]
    topA = [bottomA[i] + A[i][3] - 1 for i in range(len(A))]

    leftB = [b[0] for b in B]
    bottomB = [b[1] for b in B]
    rightB = [leftB[i] + B[i][2] - 1 for i in range(len(B))]
    topB = [bottomB[i] + B[i][3] - 1 for i in range(len(B))]

    overlap = []
    for i in range(len(leftA)):
        tmp = (max(0, min(rightA[i], rightB[i]) - max(leftA[i], leftB[i])+1)
            * max(0, min(topA[i], topB[i]) - max(bottomA[i], bottomB[i])+1))
        areaA = A[i][2] * A[i][3]
        areaB = B[i][2] * B[i][3]
        overlap.append(tmp/float(areaA+areaB-tmp))

    return overlap
